Adds data-exclude support and localStorage.plausible_ignore support

Vigasaurus opened the pull request:
### Changes

Adds support in the plausible.js script for exlcuding specific routes from couting towards events, as well as preventing all plausible events if a localStorage flag `plausible_ignore` is present. 

@ukutaht This isn't *technically* done yet, I wanted your feedback on methodology and usability before finalizing it. I also am mostly unsure what the `p.js` script is for/when its used (I found no meaningful reference to it) - so I'm unsure if my changes need to be added to it. 

Once we can agree on the format/the method - I can write up the docs/changelog as needed. 

Basic premise is that we have two new exclusions on the client-side, similar to localhost or phantom. 
First, `localStorage.plausible_ignore` - if this exists at all (i.e. is truthy in its string representation), all plausible events are prevented. 
Second, if a string is passed in to `data-exclude` with the format `data-exclude="/tou, /rule/*, /how-to-play, /*/admin, */secret, */priv/*"` pages that match those routes will be excluded. In this example, we read those routes as the following table. I decided to use a format like this in pseudo-regex to make it a bit more usable/easy to understand for an end user - lmk if you'd prefer just a straight regex or something.

| data-exclude input | parsed RegEx | 
| ------------- | ------------- | 
| `/tou` | `/^\/tou\/?$/` | 
| `/rule/*` | `/^\/rule\/.*\/?$/` | 
| `/how-to-play` | `/^\/how-to-play\/?$/` | 
| `/*/admin` | `/^\/.*\/admin\/?$/` | 
| `*/secret` | `/^.*\/secret\/?$/` | 
| `*/priv/*` | `/^.*\/priv\/.*\/?$/` | 

Also, a sidenote - it seems like the plausible script has creeped above 1KB - sitting at 1268B (1.24 KiB = 1.27 KB) on master - and with this update it creeps up to 1591B (1.55 KiB = 1.59 KB). I don't think it super matters, since we are adding features to get to this size - but I just figure some branding/copy on the landing might need some updating to match it. 

### Tests
- [X] This PR does not require tests

### Changelog
- [X] Entry has been added to changelog

### Documentation
- [X] [Docs](https://github.com/plausible/docs) have been updated
https://github.com/plausible/docs/pull/45

ukutaht commented:
Looks good overall @Vigasaurus 

`p.js` is deprecated but kept around to support some early users. No need to worry about it. If someone using that script wants access to new features, they are welcome to use the new `plausible.js` script.

>  First, localStorage.plausible_ignore - if this exists at all (i.e. is truthy in its string representation), all plausible events are prevented.

I haven't tested yet but my only worry is someone doing `localStorage.plausible_ignore = false` to disable it and getting confused because that evaluates to a truthy value in our script? Not sure how localStorage serialization works but setting `localStorage.plausible_ignore = false` should make the script count again.

> I decided to use a format like this in pseudo-regex to make it a bit more usable/easy to understand for an end user - lmk if you'd prefer just a straight regex or something.

I think it's great, much preferred to regular Regex syntax and I think using stars in URLs is very widely accepted/supported.

> Also, a sidenote - it seems like the plausible script has creeped above 1KB - sitting at 1268B (1.24 KiB = 1.27 KB) on master - and with this update it creeps up to 1591B (1.55 KiB = 1.59 KB). I don't think it super matters, since we are adding features to get to this size - but I just figure some branding/copy on the landing might need some updating to match it.

This is the raw size but we advertise the size with Brotli compression which is what you actually get over the wire. With compression it's probably not over 1KB yet.

Since adding page exclusions requires one to re-deploy their script anyways, I think we should add it as an extra script option that doesn't get included in the standard script. See how I've done `hashMode` and `outboundLinks`.

Vigasaurus commented:
Coolio, thanks for the detailed feedback on all the PRs :D I'll get going on all the comments some time today or tomorrow 😀

Vigasaurus commented:
@ukutaht This should be ready to go - I'm happy with most everything here, but I'm unsure about the formatting in `tracker.ex` and the very long array in the map - do lmk if you'd prefer some different formatting there.

ukutaht commented:
Will review later today but wanted to jot down a thought so I'll remember to test later on.

What's the behaviour of the pseudo-regex with trailing slashes and subdirectories? What I mean is:
| Exclusion rule  | Actual URL | Excluded? |
| ------------------- | --------------- | -------------- |
| /page    | /page  | Yes
| /page  | /page/  | Yes (?)
| /page/ | /page/ | Yes
| /page/ | /page | No (?)
| /page | /page/nested | No
| /page/ | /page/nested | No
| /page/* | /page/nested | Yes
| /page/* | /page | No (?)
| /page/* | /page/ | No (?)

The 'Excluded' column is what I intuitively expect to happen, question marks mean that I am not 100% certain this is the expected/correct behaviour. However it is implemented, it should be well documented and as easy to understand as possible.

Vigasaurus commented:
Yeah this is definitely a sticking point for me too, as far as documentation is concerned. I feel as though a technical explanation of what the regex becomes would be best for power-users, but would likely confuse others. I think I kept it pretty well explained in https://github.com/plausible/docs/pull/45 but it could definitely use more detail if you think its unclear. (Maybe this table you've laid out can help actually)

Exclusion rule | Actual URL | Excluded? | Actually Excluded?
-- | -- | -- | --
/page | /page | Yes | Yes
/page | /page/ | Yes (?) | Yes
/page/ | /page/ | Yes | Yes
/page/ | /page | No (?) | No
/page | /page/nested | No | No
/page/ | /page/nested | No | No
/page/* | /page/nested | Yes | Yes
/page/* | /page | No (?) | No
/page/* | /page/ | No (?) | Yes (!)

Same table as yours, last column is whether or not the current implementation excludes it

So most of them work how you'd expect - the last one is the only one which doesn't and that's mostly just a decision to be made here. Do we allow zero-length `*`'s or no - I'd say yes, but up to you (it's a single-character fix if not).  
Generally, I'd say that the trailing slash be fully optional like it is right now is best - since some browsers always add it, some always strip it, and others leave it the same - so I think making it always an option makes it most likely to match when it should, but again, up to you. 

The way I tested these was by just extracting the regex constructor from the script and running it manually. Here's what it became if you want to try it at all. 
```javascript
"actual_pathname_to_test".match(new RegExp('^' + "excluded-string".trim().replace(/\*/g, '.*') + '\/?$'))
```

so I tested your examples by doing for example
```javascript
"/page/nested".match(new RegExp('^' + "/page/*".trim().replace(/\*/g, '.*') + '\/?$'))
```

Vigasaurus commented:
I'm actually going to make a change here (tomorrow, that's why I'm moving to draft) to make docs and explaining it simpler, in that the `*` in the provided pattern can't match a `/`, such that `/page/*` will only match `/page/nested/` and not `/page/nested/doublenested/etc/`. I think this is a more consistent implementation with what stars in URLs usually mean elsewhere. 

If this format all works well etc, I'll eventually add this same matching to page goals if wanted (for stuff like WooCommerce's `/order/confirmation/<order_id>`

Vigasaurus commented:
This is now updated to not let the `*` match `/`s or whitespace. I'm updating the docs setup still, and will get some updated examples here as well. Once both are done I'll move this off draft, but it's otherwise done.

Vigasaurus commented:
> Honestly, I think the biggest confusion point will be trailing slashes

I added a note in the docs saying that trailing slashes are added automatically and are not needed in date-exclude.

> Double Nesting

I've changed it such that a `*` in data-exclude cannot match a `/` or a whitespace character in the pathname, as such paths such as `/admin/*` now only matches `/admin/1` and similar, but not `/admin/1/edit` nor `/admin`

> the last one is the only one which doesn't and that's mostly just a decision to be made here. Do we allow zero-length `*`'s or no - I'd say yes, but up to you (it's a single-character fix if not).

More research found that zero-length asterisks in URLs in this format are non-standard, so I've changed them to be 1-length or greater. 

Updated page path testing regex: 
> ```js
> "actual_pathname_to_test".match(new RegExp('^' + "split-excluded-string".trim().replace(/\*/g, '[^\\s\/]+') + '\/?$'))
> ```

Overall, I added much more detail in plausible/docs#45, and an option for tables in the docs. 


Please do let me know your thoughts :D

ukutaht commented:
Again sorry about the super long wait on this :)

Overall this is in a great place. Good job on the documentation side as well.

> I've changed it such that a * in data-exclude cannot match a / or a whitespace character in the pathname, as such paths such as /admin/* now only matches /admin/1 and similar, but not /admin/1/edit nor /admin

Sounds good. I think we should use bash glob as the main inspiration for the syntax here. It's well known. In glob, a regular asterisk does not match a forward slash as you've done here. But I think we should add a way to exclude a whole subdirectory.

In glob, this can be done with `/admin/**`. What do you think about adding `**` as a way to match any character, including forward slash?

Also in glob both single and double asterisks do match a zero-length character. So in glob `/page*` matches with `/page` whereas in this implementation it doesn't seem to. Is there a good reason to require a character when `*` is specified? 

> If this format all works well etc, I'll eventually add this same matching to page goals if wanted (for stuff like WooCommerce's /order/confirmation/<order_id>

Love it. The same syntax could also be used to roll up pageviews in regular reporting (i.e. in the Top Pages report). I know many people want to group pageviews with similar URLs.

Vigasaurus commented:
Love the ideas. I think the double glob mechanic would definitely be useful too (and should be easy to implement, since it's just the old implementation tacked on). 

I have no particular issues with either implementation as far as zero-length goes - I think it just comes down to consistency and documentation. 


I'll make the changes 👍
- Zero Length is allowed
- Double Glob
- Docs update for the above

ukutaht commented on tracker/src/plausible.js:
```diff
@@ -6,11 +6,22 @@
     if (/^localhost$|^127(?:\.[0-9]+){0,2}\.[0-9]+$|^(?:0*\:)*?:?0*1$/.test(location.hostname) || location.protocol === 'file:') return console.warn('Ignoring event on localhost');
     if (window.phantom || window._phantom || window.__nightmare || window.navigator.webdriver) return;
+    {{#if exclusionMode}}
+    if (plausible_ignore && JSON.parse(plausible_ignore)) return console.warn("Ignoring event due to localStorage flag")
```
Could probably save a few bytes in the generated code by doing `plausible_ignore == "true"` instead without the presence check. I know parsing as JSON is more *proper/correct* but in this script we are counting every byte that goes over the wire.

Second thing you can try is helping the minifier reduce the duplication between the calls to `console.warn`. You can check if extracting a function like:

```javascript
function ignore(reason) {
  console.warn('Ignoring event: ' + reason)
}
```
and in the two early returns you can use `return ignore('localStorage flag')`

These static strings are a big part of the payload because they don't minify well.

ukutaht reviewed (changes requested):
Looks good, requested a few changes. Once they're done I will merge this

ukutaht commented on lib/plausible_web/plugs/tracker.ex:
```diff
@@ -4,14 +4,25 @@ defmodule PlausibleWeb.Tracker do
+    "plausible.hash.outbound-links.js" => ["plausible.outbound-links.hash.js"],
+    "plausible.hash.exclusions.js" => ["plausible.exclusions.hash.js"],
+    "plausible.exclusions.outbound-links.js" => ["plausible.outbound-links.exclusions.js"],
+    "plausible.hash.exclusions.outbound-links.js" => [
```
To clarify, not as part of this PR.

metmarkosaric commented:
Got a question on Twitter: they're using the same script on both the main domain and subdomain but want to exclude the `/` so the pages doesn't get counted twice. Any thoughts?

https://twitter.com/tim_nolet/status/1354040020018077698

Vigasaurus commented:
I'm a tiny bit confused by the use case/end goal, but can confirm that "/**" would exclude any path

metmarkosaric commented:
thanks @Vigasaurus! i've asked them to clarify. btw if you exclude a specific page does that mean that any custom goals on that specific page won't be counting?

Vigasaurus commented:
Nope, the exclusion only happens for triggering pageviews - outbound links and custom events still fire.

Vigasaurus commented:
Also reading more of the Twitter thread, the use case makes sense to me. I think excluding "/**" is definitely the way to do it for them, if they want all the marketing pages on the www or none subdomain, and then in app they run the full exclusion and fire custom events as needed.

metmarkosaric commented:
thanks! so by excluding "/**" no visitors / pageviews will be counted both from the main domain and from the subdomain?

Vigasaurus commented:
Well, I assume they have separate html outputs for each subdomain, so they'd likely only put the exclude on one of the subs - the main domain (the marketing site) should just not have any exclusion

metmarkosaric commented:
ahh yes! i see. makes sense to me now. thanks!

Vigasaurus commented:
Sweet! Feel free to ping me here or on the Twitter thread if they have other concerns/questions :D

metmarkosaric commented:
they responded with more details so i think it's more clear now: https://twitter.com/tim_nolet/status/1354048423742341120

last question: if they do as described will the referral source still show for custom event conversions?

Vigasaurus commented:
Mm, I don't think so, because from my understanding the referral source for custom events comes by linking the entry page+referrer for the entire session to the event session, and if there are no pageviews there are no entry pages or referrers. But to be honest I haven't looked a whole lot at how the referrer and sessions stuff works, so Uku would know better/more concretely.

metmarkosaric commented:
ok cool, thanks!

ukutaht commented:
I think the referrer stuff and everything should work normally. In their setup the normal funnel seems to:
1. Generate some pageviews on www.domain.com
2. Signup on app.domain.com

As long as they're sending the stats to the same dashboard, the session should stay intact and the referrer source for the initial pageview on www.domain.com will be visible in the dashboard.

metmarkosaric commented:
perfect! thanks for confirming!

Vigasaurus commented:
Oh yeah that makes sense! I guess the generated user hash would line up the same too since the data-domain would be the same. 

I guess the break down would happen if someone was referred directly to app.domain.com/signup, and that wouldn't get any referrer on load, but that's probably a bit unlikely/could be planned around from a subdomain design standpoint

ukutaht commented:
A (potential) bug raised by one of our clients:
Exclusion rule: `/l/*`
Actual page path: `/l/some-page-here`

Expected to match but this is the result:
```js
"/l/testing".match(new RegExp("/l/*".trim().replace(/\*\*/g, '.*').replace(/[^\.]\*/g, '[^\\s\/]*') + '\/?$'))
<- null
```
@Vigasaurus any idea what's going on here?

Vigasaurus commented:
@ukutaht Ah yeah that's actually a pretty big bug, that I caught when I was doing pageview goal globs, but completely forgot to update in the tracker script. Notice [here](https://github.com/plausible/analytics/blob/master/lib/plausible/stats/clickhouse.ex#L936) the regex for the second replace was appropriately changed to a negative lookbehind, instead of the negative replacement, but in the tracker I didn't make this change. The appropriate new version in the tracker would be:

```javascript
"/l/testing".match(new RegExp("/l/*".trim().replace(/\*\*/g, '.*').replace(/(?<!\.)\*/g, '[^\\s\/]*') + '\/?$'))
<- Array [ "/l/testing" ]
```

which is the correct result, notice the swap from `.replace(/[^\.]\*/g, '[^\\s\/]*')` to `.replace(/(?<!\.)\*/g, '[^\\s\/]*')`. Will PR shortly.
